File: 프로그래머스/Lv2/두 큐 합 같게 만들기(X).py
# 다른 사람 풀이
def solution(queue1, queue2):
    target = (sum(queue1) + sum(queue2)) // 2
    cur = sum(queue1)
    queue3 = queue1 + queue2 + queue1

    s = 0
    e = len(queue1) - 1
    answer = 0
    while True:
        if cur == target:
            return answer
        if cur < target:
            e += 1
            if e >= len(queue3):
                return -1
            cur += queue3[e]
        else:
            cur -= queue3[s]
            s += 1
        answer += 1

print(solution([3, 2, 7, 2], [4, 6, 5, 1]))
print(solution([1, 2, 1, 2], [1, 10, 1, 2]))
print(solution([1, 1], [1, 5]))

File: 프로그래머스/Lv2/두 큐 합 같게 만들기(X).js
// 내 풀이 ( 시간 초과 )
function solution(q1, q2) {
    let answer = 0;
    let sum1 = q1.reduce((a,c) => a+c, 0);
    let sum2 = q2.reduce((a,c) => a+c, 0);
    
    while(q1.length && q2.length){
        if(sum1 > sum2){
            let temp = q1.shift();
            q2.push(temp);
            sum1 -= temp;
            sum2 += temp;
            answer++;
        }else if(sum1 < sum2){
            let temp = q2.shift();
            q1.push(temp);
            sum1 += temp;
            sum2 -= temp;
            answer++;
        }

        if(sum1 === sum2){
            return answer;
        }
    }
    
    return -1;
}

// 다른 사람 풀이
const solution2 = (queue1, queue2) => {
    let sum1 = queue1.reduce((prev, cur) => prev + cur, 0);
    let sum2 = queue2.reduce((prev, cur) => prev + cur, 0);
    const half = (sum1 + sum2) / 2;
    const q = [...queue1, ...queue2];
    let q1Pointer = 0;
    let q2Pointer = queue1.length;
  
    for (let cnt = 0; cnt < queue1.length * 3; cnt++) {
      if (sum1 === half) {
        return cnt;
      }
      sum1 =
        sum1 > half
          ? sum1 - q[q1Pointer++ % q.length]
          : sum1 + q[q2Pointer++ % q.length];
    }
  
    return -1;
  };